### PR TITLE
executor: implement json time/duration

### LINKED ISSUE #9988: JSON data type does not support custom types (i.e. MySQL types)
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```SQL
select json_type(cast(cast('2014-11-25 18:00' as datetime) as json));
select json_type(cast(cast('2014-11-25' as date) as json));
select json_type(cast(cast('18:00:59' as time) as json));
```

2. What did you expect to see?
## In MySQL
```SQL
mysql> select json_type(cast(cast('2014-11-25 18:00' as datetime) as json));
+---------------------------------------------------------------+
| json_type(cast(cast('2014-11-25 18:00' as datetime) as json)) |
+---------------------------------------------------------------+
| DATETIME                                                      |
+---------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select json_type(cast(cast('2014-11-25' as date) as json));
+-----------------------------------------------------+
| json_type(cast(cast('2014-11-25' as date) as json)) |
+-----------------------------------------------------+
| DATE                                                |
+-----------------------------------------------------+
1 row in set (0.00 sec)

mysql> select json_type(cast(cast('18:00:59' as time) as json));
+---------------------------------------------------+
| json_type(cast(cast('18:00:59' as time) as json)) |
+---------------------------------------------------+
| TIME                                              |
+---------------------------------------------------+
1 row in set (0.01 sec)

mysql> 
mysql> select version();
+-------------------------+
| version()               |
+-------------------------+
| 5.7.25-0ubuntu0.18.04.2 |
+-------------------------+
1 row in set (0.00 sec)

mysql> 
```

3. What did you see instead?

## In TiDB
```SQL

mysql> select json_type(cast(cast('2014-11-25 18:00' as datetime) as json));
+---------------------------------------------------------------+
| json_type(cast(cast('2014-11-25 18:00' as datetime) as json)) |
+---------------------------------------------------------------+
| STRING                                                        |
+---------------------------------------------------------------+
1 row in set (0.01 sec)

mysql> select json_type(cast(cast('2014-11-25' as date) as json));
+-----------------------------------------------------+
| json_type(cast(cast('2014-11-25' as date) as json)) |
+-----------------------------------------------------+
| STRING                                              |
+-----------------------------------------------------+
1 row in set (0.00 sec)

mysql> select json_type(cast(cast('18:00:59' as time) as json));
+---------------------------------------------------+
| json_type(cast(cast('18:00:59' as time) as json)) |
+---------------------------------------------------+
| STRING                                            |
+---------------------------------------------------+
1 row in set (0.00 sec)

mysql> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v3.0.0-beta-321-gc6a757b24
Git Commit Hash: c6a757b2435965a72ada9d61ea8b7a367bd89b5f
Git Branch: master
UTC Build Time: 2019-04-02 02:38:04
GoVersion: go version go1.12.1 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
1 row in set (0.00 sec)

mysql> 
```

## SIG slack channel

 [#sig-exec](https://slack.tidb.io/invite?team=tidb-community&channel=sig-exec&ref=high-performance)

## Score

300

## Mentor

- @lzmhhh123

### LINKED ISSUE #31104: In the test cases in TiDB, MySQL and TiDB behave differently, whichever prevails
## General Question

In the test case  `util/admin/admin_integration_test.go:59`
```
create table t1 (
		a int             as (JSON_EXTRACT(k,'$.a')),
		c double          as (JSON_EXTRACT(k,'$.c')),
		d decimal(20,10)  as (JSON_EXTRACT(k,'$.d')),
		e char(10)        as (JSON_EXTRACT(k,'$.e')),
		f date            as (JSON_EXTRACT(k,'$.f')),
		g time            as (JSON_EXTRACT(k,'$.g')),
		h datetime        as (JSON_EXTRACT(k,'$.h')),
		i timestamp       as (JSON_EXTRACT(k,'$.i')),
		j year            as (JSON_EXTRACT(k,'$.j')),
		k json);

insert into t1 set k='{"a": 100,"c":1.234,"d":1.2340000000,"e":"abcdefg","f":"2018-09-28","g":"12:59:59","h":"2018-09-28 12:59:59","i":"2018-09-28 16:40:33","j":"2018"}';
```

In TiDB it can write normally。

When executed in MySQL
```sql
mysql> insert into t1 set k='{"a": 100,"c":1.234,"d":1.2340000000,"e":"abcdefg","f":"2018-09-28","g":"12:59:59","h":"2018-09-28 12:59:59","i":"2018-09-28 16:40:33","j":"2018"}';
ERROR 3156 (22018): Invalid JSON value for CAST to DATE/TIME/DATETIME/TIMESTAMP from column json_extract at row 1
```

Which shall prevail

### PR DESCRIPTION
**Pull request by YangKeao**
Signed-off-by: YangKeao <yangkeao@chunibyo.icu>

### What problem does this PR solve?

Issue Number: close #9988, close #31104

Problem Summary:

The TiDB doesn't distinguish the `DATETIME/DATE/TIMESTAMP/TIME` and `STRING`.

### What is changed and how it works?

Based on https://github.com/pingcap/tidb/pull/37578

I added four types of JSON: `DATETIME/DATE/TIMESTAMP/TIME`, and the cast between them and the `Time/Duration`.

### Check List

Tests 

TODO

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

```release-note
Fix the issue that TiDB doesn't distinguish the `DATETIME/DATE/TIMESTAMP/TIME` and `STRING` in json type.
```

### CODE REVIEW

**Comment by YangKeao** [comment]
PTAL @xiongjiwei

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -1740,20 +1740,34 @@ func (b *builtinCastJSONAsTimeSig) evalTime(row chunk.Row) (res types.Time, isNu
...
+			ts = gotime.Now()
+		}
+		res, err = duration.ConvertToTimeWithTimestamp(sc, b.tp.GetType(), ts)
+		if err != nil {
+			return types.ZeroTime, true, handleInvalidTimeError(b.ctx, err)
+		}
+		res, err = res.RoundFrac(sc, b.tp.GetDecimal())
+		return res, isNull, err
+	default:
+		err = types.ErrTruncatedWrongVal.GenWithStackByArgs(types.TypeStr(b.tp.GetType()), val.String())
```
Comment (xiongjiwei) [inline expression/builtin_cast.go:1768]: does it block string to time? e.g.
```sql
create table t(a json);
insert into t value (json_object('a', '2020-09-02 00:00:00'));
select cast(a->'$.a' as datetime) from t;
```

Comment (xiongjiwei) [inline expression/builtin_cast.go:1768]: vec version will `handleInvalidTimeError`, does the non-vec need it?

File: types/datum_test.go [hunk 2]
```diff
@@ -225,12 +225,17 @@ func TestConvertToFloat(t *testing.T) {
...
-func mustParseTimeIntoDatum(s string, tp byte, fsp int) (d Datum) {
+func mustParseTime(s string, tp byte, fsp int) Time {
 	t, err := ParseTime(&stmtctx.StatementContext{TimeZone: time.UTC}, s, tp, fsp)
 	if err != nil {
 		panic("ParseTime fail")
 	}
+	return t
+}
+
+// mustParseTimeIntoDatum is similar to ParseTime but panic if any error occurs.
```
Comment (xiongjiwei) [inline types/datum_test.go:236]: seems this comment can be remove

File: expression/integration_test.go [hunk 3]
```diff
@@ -7523,3 +7523,22 @@ func TestCastRealAsTime(t *testing.T) {
...
+	tk.MustExec("insert into t values (1, CAST('1998-06-13 12:12:12' as DATETIME))")
+	tk.MustExec("insert into t values (2, DATE('1596-03-31'))")
+	tk.MustExec("insert into t values (3, CAST('1596-03-31 12:12:12' as DATETIME))")
+	tk.MustQuery("select i, cast(j as date), cast(j as datetime), cast(j as time) from t").Check(testkit.Rows(
+		"0 1998-06-13 1998-06-13 00:00:00 00:00:00",
+		"1 1998-06-13 1998-06-13 12:12:12 12:12:12",
+		"2 1596-03-31 1596-03-31 00:00:00 00:00:00",
+		"3 1596-03-31 1596-03-31 12:12:12 12:12:12",
+	))
+}
```
Comment (xiongjiwei) [inline expression/integration_test.go:7544]: can you please also add a test like `select json_type(j) from t`?

File: expression/builtin_cast.go (hunk 1 above)
Comment (xiongjiwei) [reply expression/builtin_cast.go:1768]: mysql8.0 gives results, we intend to be compatible with 8.0

Comment (YangKeao) [reply expression/builtin_cast.go:1768]: Good. I'll add another branch to convert json string to time/duration.

Comment (YangKeao) [reply expression/builtin_cast.go:1783]: Fixed. And also added a test for it :smiley:

Comment (YangKeao) [reply expression/builtin_cast.go:1783]: Yes. It's a `ErrTruncatedWrongVal`, so I should pass it to `b.ctx.GetSessionVars().StmtCtx.HandleTruncate` directly.

Comment (YangKeao) [reply expression/builtin_cast.go:1783]: Fixed.

File: types/datum_test.go (hunk 2 above)
Comment (YangKeao) [reply types/datum_test.go:236]: Why? This comment is for the next function, and that function is still used by the test :thinking: .

File: expression/integration_test.go (hunk 3 above)
Comment (YangKeao) [reply expression/integration_test.go:7552]: Yes!

File: types/datum_test.go (hunk 2 above)
Comment (xiongjiwei) [reply types/datum_test.go:236]: Get it

**Review by xiongjiwei — APPROVED** [review]

**Comment by YangKeao** [comment]
@xhebox PTAL

**Review by xhebox — APPROVED** [review]

**Comment by xhebox** [comment]
@YangKeao  Some tests may be related

```
  ✘  https://revive.run/r#exported  exported var JSONZero should have comment or be unexported  
```
```
新的故障 - 2
TestIssue1223 – expression2s
TestToBool – types
```

**Comment by YangKeao** [comment]
> @YangKeao Some tests may be related
> 
> ```
>   ✘  https://revive.run/r#exported  exported var JSONZero should have comment or be unexported  
> ```
> 
> ```
> 新的故障 - 2
> TestIssue1223 – expression2s
> TestToBool – types
> ```

Fixed :beers:

**Pull request merged**